hail/python/hailtop/saige/io.py

```diff
@@ -0,0 +1,307 @@
+
+    @staticmethod
+    def from_input_files(b: hb.Batch, bfile_root: str) -> 'PlinkResourceGroup':
+        inputs = PlinkResourceGroup.inputs(bfile_root)
+        return PlinkResourceGroup._from_input_files(b, inputs)
+
+    @staticmethod
+    def from_job_intermediate(j: hb.Job) -> 'PlinkResourceGroup':
+        inputs = PlinkResourceGroup.inputs('{root}')
+        return PlinkResourceGroup._from_job_intermediate(j, inputs)
+
+
+class VCFResourceGroup(BaseSaigeResource):
+    @staticmethod
+    def inputs(root: str) -> ResourceGroupInputs:
+        return ResourceGroupInputs('vcf', vcf=f'{root}.vcf.gz', tbi=f'{root}.vcf.gz.tbi')
+
+    @staticmethod
+    async def use_checkpoint_if_exists_and_requested(
+        fs: AsyncFS, pool: AsyncWorkerPool, b: hb.Batch, config: CheckpointConfigMixin, vcf_root: str
+    ) -> Optional['VCFResourceGroup']:
+        inputs = VCFResourceGroup.inputs(vcf_root)
+        return await VCFResourceGroup._use_checkpoint_if_exists_and_requested(fs, pool, b, config, inputs)
+
+    @staticmethod
+    def from_input_files(b: hb.Batch, root: str) -> 'VCFResourceGroup':
+        inputs = VCFResourceGroup.inputs(root)
+        return VCFResourceGroup._from_input_files(b, inputs)
+
+    @staticmethod
+    def from_job_intermediate(j: hb.Job) -> 'VCFResourceGroup':
+        inputs = VCFResourceGroup.inputs('{root}')
+        return VCFResourceGroup._from_job_intermediate(j, inputs)
+
+
+class BgenResourceGroup(BaseSaigeResource):
+    @staticmethod
+    def inputs(root: str) -> ResourceGroupInputs:
+        return ResourceGroupInputs('bgen', bgen=f'{root}.bgen', bgi=f'{root}.bgen.bgi', sample=f'{root}.sample')
+
+    @staticmethod
+    async def use_checkpoint_if_exists_and_requested(
+        fs: AsyncFS, pool: AsyncWorkerPool, b: hb.Batch, config: CheckpointConfigMixin, bgen_root: str
+    ) -> Optional['BgenResourceGroup']:
+        inputs = BgenResourceGroup.inputs(bgen_root)
+        return await BgenResourceGroup._use_checkpoint_if_exists_and_requested(fs, pool, b, config, inputs)
+
+    @staticmethod
+    def from_input_files(b: hb.Batch, root: str) -> 'BgenResourceGroup':
+        inputs = BgenResourceGroup.inputs(root)
+        return BgenResourceGroup._from_input_files(b, inputs)
+
+    @staticmethod
+    def from_job_intermediate(j: hb.Job) -> 'BgenResourceGroup':
+        inputs = BgenResourceGroup.inputs('{root}')
+        return BgenResourceGroup._from_job_intermediate(j, inputs)
+
+
+class SaigeSparseGRMResourceGroup(BaseSaigeResource):
+    @staticmethod
+    def inputs(root: str) -> ResourceGroupInputs:
+        return ResourceGroupInputs('grm', grm=f'{root}.sparseGRM.mtx', sample_ids=f'{root}.sparseGRM.mtx.sampleIDs.txt')
+
+    @staticmethod
+    async def use_checkpoint_if_exists_and_requested(
+        fs: AsyncFS, pool: AsyncWorkerPool, b: hb.Batch, config: CheckpointConfigMixin, grm_root: str
+    ) -> Optional['SaigeSparseGRMResourceGroup']:
```

Comment:
Actually, I can probably consolidate this even more with more classmethods. Will do that in the next round.